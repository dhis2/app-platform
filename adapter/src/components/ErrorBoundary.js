import i18n from '@dhis2/d2-i18n'
import cx from 'classnames'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import buttonStyles from './styles/Button.style'
import styles from './styles/ErrorBoundary.style'

// In order to avoid using @dhis2/ui components in the error boundary - as anything
// that breaks within it will not be caught properly - we define a component
// with the same styles as Button
const UIButton = ({ children, onClick }) => (
    <>
        <style jsx>{buttonStyles}</style>
        <button onClick={onClick}>{children}</button>
    </>
)

UIButton.propTypes = {
    children: PropTypes.node.isRequired,
    onClick: PropTypes.func.isRequired,
}

const translatedErrorHeading = i18n.t(
    'An error occurred in the DHIS2 application.'
)

export class ErrorBoundary extends Component {
    constructor(props) {
        super(props)
        this.state = {
            error: null,
            errorInfo: null,
            drawerOpen: false,
        }
        this.errorDetailsRef = React.createRef()
    }

    componentDidCatch(error, errorInfo) {
        this.setState({
            error,
            errorInfo,
        })
    }

    toggleTechInfoDrawer = () => {
        this.setState({
            drawerOpen: !this.state.drawerOpen,
        })
    }

    handleCopyErrorDetails = () => {
        const errorDetails = this.errorDetailsRef.current.textContent
        navigator.clipboard.writeText(errorDetails).then(() => {
            alert(i18n.t('Technical details copied to clipboard'))
        })
    }

    render() {
        const { children, fullscreen, onRetry } = this.props
        if (this.state.error) {
            return (
                <div className={cx('mask', { fullscreen })}>
                    <style jsx>{styles}</style>
                    <div className="container">
                        <h1 className="message">
                            {i18n.t('Something went wrong')}
                        </h1>
                        {onRetry && (
                            <div className="retry">
                                <UIButton onClick={onRetry}>
                                    {i18n.t('Try again')}
                                </UIButton>
                            </div>
                        )}
                        <button
                            className="drawerToggle"
                            onClick={this.toggleTechInfoDrawer}
                        >
                            {this.state.drawerOpen
                                ? i18n.t('Hide technical details')
                                : i18n.t('Show technical details')}
                        </button>
                        <div
                            className={cx('drawer', {
                                hidden: !this.state.drawerOpen,
                            })}
                        >
                            <div className="errorIntro">
                                <p>{translatedErrorHeading}</p>
                                <p>
                                    {i18n.t(
                                        'The following information may be requested by technical support.'
                                    )}
                                </p>
                                <UIButton onClick={this.handleCopyErrorDetails}>
                                    {i18n.t(
                                        'Copy technical details to clipboard'
                                    )}
                                </UIButton>
                            </div>
                            <pre
                                className="errorDetails"
                                ref={this.errorDetailsRef}
                            >
                                {`${this.state.error}\n`}
                                {this.state.error.stack}
                                {this.state.errorInfo.componentStack}
                            </pre>
                        </div>
                    </div>
                </div>
            )
        }

        return children
    }
}

ErrorBoundary.propTypes = {
    children: PropTypes.node.isRequired,
    fullscreen: PropTypes.bool,
    onRetry: PropTypes.func,
}
