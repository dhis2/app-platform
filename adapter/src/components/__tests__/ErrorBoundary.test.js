import { shallow } from 'enzyme'
import React from 'react'
import { ErrorBoundary } from '../ErrorBoundary'

const Something = () => {
    // Placeholder
    return null
}

describe('ErrorBoundary', () => {
    it('Should render the normal tree when no error occurs', () => {
        const wrapper = shallow(
            <ErrorBoundary classes={{}}>
                <div>
                    <span id="testme">Hello there!</span>
                </div>
            </ErrorBoundary>
        )

        expect(wrapper.find('span#testme').length).toBe(1)
    })

    it('Should render the error mask when an error is thrown', () => {
        const wrapper = shallow(
            <ErrorBoundary classes={{}}>
                <Something />
            </ErrorBoundary>
        )

        expect(wrapper.find(Something).length).toBe(1)

        const testErrorText =
            'This is a test of the emergency alert system.  This is only a test.'
        wrapper.find(Something).simulateError(new Error(testErrorText))
        expect(wrapper.contains('Something went wrong')).toBe(true)
    })

    it('Should match the snapshot tree when error is manually invoked', () => {
        const wrapper = shallow(
            <ErrorBoundary classes={{}}>
                <Something />
            </ErrorBoundary>
        )

        const error = {
            stack: 'This is the stack',
        }
        const errorInfo = {
            componentStack: 'This is the component stack',
        }

        wrapper.setState({ error, errorInfo })
        expect(wrapper.contains('Something went wrong')).toBe(true)

        expect(wrapper).toMatchSnapshot()
    })
})
