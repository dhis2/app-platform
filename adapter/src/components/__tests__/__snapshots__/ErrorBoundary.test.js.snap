// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary Should match the snapshot tree when error is manually invoked 1`] = `ShallowWrapper {}`;
